File: java/src/main/java/workshop/encryption/exercise2/Asymmetric.java
package workshop.encryption.exercise2;

import helpers.RandomNonce;

import org.abstractj.kalium.crypto.Box;
import org.abstractj.kalium.keys.PrivateKey;
import org.abstractj.kalium.keys.PublicKey;

public class Asymmetric {
    
    public static byte[] encrypt(PublicKey receiverPublicKey,
            PrivateKey senderPrivateKey, RandomNonce nonce, byte[] message) {
        Box box = new Box(receiverPublicKey, senderPrivateKey);
        return box.encrypt(nonce.getBytes(), message);
    }
    
    public static byte[] decrypt(PrivateKey receiverPrivateKey,
            PublicKey senderPublicKey, RandomNonce nonce, byte[] encryptedMessage) {
        Box box = new Box(senderPublicKey, receiverPrivateKey);
        return box.decrypt(nonce.getBytes(), encryptedMessage);
    }
}

File: ruby/lib/exercise2/asymmetric.rb
module Asymmetric

  def self.encrypt(receiver_public_key, sender_private_key, message)
    box = RbNaCl::SecretBox.encrypt(receiver_public_key, sender_private_key)
    box.encrypt(message)
  end

  def self.decrypt(receiver_private_key, sender_public_key, ciphertext)
    box = RbNaCl::SecretBox.decrypt(sender_public_key, receiver_private_key)
    box.decrypt(ciphertext)
  end

end
